File: sdk/python/ggid/passkey.py
"""Passkey/WebAuthn registration and authentication API calls.

Usage::

    from ggid.passkey import PasskeyClient
    pk = PasskeyClient("https://ggid.example.com")
    options = pk.begin_registration(token, "my-passkey")
"""

from __future__ import annotations
import json
from typing import Any
import urllib.request


class PasskeyClient:
    """Client for WebAuthn/Passkey API calls."""

    def __init__(self, base_url: str):
        self.base_url = base_url.rstrip("/")

    def begin_registration(self, access_token: str, device_name: str) -> dict:
        """Begin WebAuthn registration. Returns server challenge options."""
        return self._post(
            "/api/v1/auth/mfa/enroll",
            {"type": "webauthn", "name": device_name},
            access_token,
        )

    def finish_registration(self, access_token: str, device_id: str, attestation: str) -> dict:
        """Finish WebAuthn registration by verifying the attestation."""
        return self._post(
            "/api/v1/auth/mfa/verify",
            {"device_id": device_id, "code": attestation},
            access_token,
        )

    def begin_login(self, username: str) -> dict:
        """Begin WebAuthn login. Returns server challenge options."""
        return self._post(
            "/api/v1/auth/webauthn/login/begin",
            {"username": username},
        )

    def finish_login(self, assertion: str) -> dict:
        """Finish WebAuthn login by verifying the assertion."""
        return self._post(
            "/api/v1/auth/webauthn/login/finish",
            {"assertion": assertion},
        )

    def _post(self, path: str, body: dict, token: str | None = None) -> dict:
        url = f"{self.base_url}{path}"
        data = json.dumps(body).encode()
        req = urllib.request.Request(url, data=data, method="POST")
        req.add_header("Content-Type", "application/json")
        if token:
            req.add_header("Authorization", f"Bearer {token}")
        with urllib.request.urlopen(req) as resp:
            return json.loads(resp.read())

File: sdk/java/src/main/java/dev/ggid/sdk/Passkey.java
package dev.ggid.sdk;

import com.google.gson.Gson;
import java.io.IOException;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.net.URI;
import java.util.Map;

/** Passkey/WebAuthn registration and authentication API calls. */
public class Passkey {
    private final Config config;
    private final HttpClient httpClient;
    private final Gson gson = new Gson();

    public Passkey(Config config, HttpClient httpClient) {
        this.config = config;
        this.httpClient = httpClient;
    }

    /** Begin WebAuthn registration. Returns server challenge options. */
    @SuppressWarnings("unchecked")
    public Map<String, Object> beginRegistration(String accessToken, String deviceName)
            throws GGIDException, IOException {
        String body = gson.toJson(Map.of("type", "webauthn", "name", deviceName));
        HttpRequest req = HttpRequest.newBuilder()
                .uri(URI.create(config.getBaseUrl() + "/api/v1/auth/mfa/enroll"))
                .header("Authorization", "Bearer " + accessToken)
                .header("Content-Type", "application/json")
                .POST(HttpRequest.BodyPublishers.ofString(body))
                .build();
        try {
            HttpResponse<String> resp = httpClient.send(req, HttpResponse.BodyHandlers.ofString());
            if (resp.statusCode() >= 400) {
                throw new GGIDException("Passkey registration failed: " + resp.body(), resp.statusCode());
            }
            return gson.fromJson(resp.body(), Map.class);
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
            throw new GGIDException("Interrupted", 0);
        }
    }

    /** Finish WebAuthn registration by verifying the attestation. */
    public void finishRegistration(String accessToken, String deviceID, String attestationResponse)
            throws GGIDException, IOException {
        String body = gson.toJson(Map.of("device_id", deviceID, "code", attestationResponse));
        HttpRequest req = HttpRequest.newBuilder()
                .uri(URI.create(config.getBaseUrl() + "/api/v1/auth/mfa/verify"))
                .header("Authorization", "Bearer " + accessToken)
                .header("Content-Type", "application/json")
                .POST(HttpRequest.BodyPublishers.ofString(body))
                .build();
        try {
            HttpResponse<String> resp = httpClient.send(req, HttpResponse.BodyHandlers.ofString());
            if (resp.statusCode() >= 400) {
                throw new GGIDException("Passkey verify failed: " + resp.body(), resp.statusCode());
            }
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
            throw new GGIDException("Interrupted", 0);
        }
    }
}
